[Backport 2.x] 13776: allow adding query parameters to RequestOptions

oliverlockwood opened the pull request:
### Description
Allows adding query parameters to RequestOptions, e.g. to be able to set `?filter-path=<something>` as documented in https://opensearch.org/docs/2.3/opensearch/common-parameters/#filtered-responses.

This is the same PR as #13777, but targeting the 2.x branch.

### Related Issues
Resolves #13776.

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

reta commented:
@oliverlockwood the backports are done automatically once https://github.com/opensearch-project/OpenSearch/pull/13777 is merged (there might be some conflicts to resolve).

reta commented:
@oliverlockwood could you please rebase against latest `2.x`? thank you.

oliverlockwood commented:
> @oliverlockwood could you please rebase against latest `2.x`? thank you.

@reta I have now rebased this off the latest `2.x`, squashed all my changes into one commit, and fixed the `CHANGELOG.md` conflict.

Please let me know if anything more is needed.  Thanks.

reta commented:
> ❌ Gradle check result for [8eb94f4](https://github.com/opensearch-project/OpenSearch/commit/8eb94f45478bde3e763625fcbd4c48a9b7036521): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39829/)

https://github.com/opensearch-project/OpenSearch/issues/13433
https://github.com/opensearch-project/OpenSearch/issues/13600

reta commented:
> ❌ Gradle check result for [8eb94f4](https://github.com/opensearch-project/OpenSearch/commit/8eb94f45478bde3e763625fcbd4c48a9b7036521): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39840/)

https://github.com/opensearch-project/OpenSearch/issues/13695

reta commented:
> ❌ Gradle check result for [8eb94f4](https://github.com/opensearch-project/OpenSearch/commit/8eb94f45478bde3e763625fcbd4c48a9b7036521): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39850/)

https://github.com/opensearch-project/OpenSearch/issues/13600
https://github.com/opensearch-project/OpenSearch/issues/11179

reta commented:
> ❌ Gradle check result for [8eb94f4](https://github.com/opensearch-project/OpenSearch/commit/8eb94f45478bde3e763625fcbd4c48a9b7036521): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39855/)

https://github.com/opensearch-project/OpenSearch/issues/13828
https://github.com/opensearch-project/OpenSearch/issues/13473

reta commented:
> ❌ Gradle check result for [8eb94f4](https://github.com/opensearch-project/OpenSearch/commit/8eb94f45478bde3e763625fcbd4c48a9b7036521): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39859/)

https://github.com/opensearch-project/OpenSearch/issues/11179
